### PR TITLE
Fix Crash due to Baking Size

### LINKED ISSUE #127: Webots crashes on startup
**Describe the Bug**

On some hardware, Webots R2019a crashes on startup.

**Screenshots**

capture

_(from ticket #324)_ 

**System**
 
It seems to affect mainly Windows users, without guarantee for now.
The bug have been reproduced with the following GPU:
  - AMD Radeon(TM) R7 M440 (0x6900)
  - AMD Radeon 6600M and 6700M Series
  - NVIDIA Corporation - GeForce 840M/PCIe/SSE2 - 3.3.0 NVIDIA 376.54
  - NVIDIA Corporation - GeForce 540M - 3.3.0

### PR DESCRIPTION
**Pull request by DavidMansolino**
Fix #127

### CODE REVIEW

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
In order to check the GPU memory, this is what I found yet:
For NVIDIA (available since OpenGL 2.0):
```
#define GL_GPU_MEM_INFO_TOTAL_AVAILABLE_MEM_NVX 0x9048

GLint nTotalMemoryInKB = 0;
glGetIntegerv(GL_GPU_MEM_INFO_TOTAL_AVAILABLE_MEM_NVX, &nTotalMemoryInKB );
```
For AMD (available since OpenGL 1.1):
```
#define GL_TEXTURE_FREE_MEMORY_ATI 0x87FC

GLint nCurAvailMemoryInKB = 0;
glGetIntegerv(GL_TEXTURE_FREE_MEMORY_ATI, &nCurAvailMemoryInKB );
```

But for AMD we only get the currently available memory and not the total memory (this is also feasible with NVIDIA). Not sure this is good.

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
Glad can easily be regenerated to include this two required define: 
http://glad.dav1d.de/#profile=core&api=gl%3D3.3&api=gles1%3Dnone&api=gles2%3Dnone&api=glsc2%3Dnone&extensions=GL_ARB_clip_control&extensions=GL_ATI_meminfo&extensions=GL_EXT_texture_filter_anisotropic&extensions=GL_NVX_gpu_memory_info&language=c&specification=gl&loader=on

But I am having issue including the `WGL_AMD_gpu_association` extension (which can be used to get the total available memory: http://nasutechtips.blogspot.com/2011/02/how-to-get-gpu-memory-size-and-usage-in.html) in glad.

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
Here are some result with the AMD Radeon R7 240 (1Gb) GPU (tested on the khepera gripper world):
  - **With all OpenGL feature disabled:** 1-1.5x
  - **Enabling anti-aliasing:** speed is divided by 2
  - **Enabling GTAO:** speed is divided by 1.5 to 3 (depending on the quality)
  - **Changing the texture resolution (except the background backing):** No visible impact
  - **Changing the HDR background resolution**: 128x128 crashes, 64x64 is working fine with all the worlds of the guided tour.

The crash with the background backing is only in the HDr case because the texture format is `RGB32F` (instead of `RGBA8`). So probably we can be more 'restrictive' in the resolution reduction in the HDR case.

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
The only remaining problem is that we had one crash report with an R7 GPU with 4Gb of memory, in that case the resolution limitation will not be applied.
I am trying to find a list of GPU device ID for the R5 and R7 series to make sure we use the resolution limitation for them, but I can't find a good list (for now I only found https://gist.github.com/roalercon/51f13a387f3754615cce but this list is not complete and doesn't even includes the one of the R7  with 4Gb of memory (0x6900)).

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
👍

**Comment by omichel** [comment]
:eyes:

File: src/wren/GlState.hpp [hunk 1]
```diff
@@ -115,6 +115,7 @@ namespace wren {
     const char *version();
     const char *glslVersion();
 
+    int gpuMemory();
```
Comment (omichel) [inline src/wren/GlState.hpp]: I would add a comment here that explain that it's 0 for Intel cards, the currently available GPU memory for AMD cards and the total GPU memory for NVIDIA cards,

File: src/webots/gui/WbView3D.cpp [hunk 2]
```diff
@@ -1225,6 +1225,29 @@ void WbView3D::checkRendererCapabilities() {
...
+    message += tr("Webots has detected that you are using an old AMD GPU. "
+                  "A recent NVIDIA or AMD graphics adapter is highly recommended to run Webots smoothly. ");
+    disableCameraAntiAliasing = true;
+    disableSMAA = true;
+    disableGTAO = true;
+    reduceTextureQuality = 1;
+  }
+#endif
+
+  // check GPU memory
```
Comment (omichel) [inline src/webots/gui/WbView3D.cpp]: I would rule out Intel cards from this test as they have 0 GPU memory, but seems to work flawlessly with any texture quality.

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
test me

**Comment by DavidMansolino** [comment]
👍

**Pull request merged**